packages/backend/src/managers/apiServer.ts

```diff
@@ -0,0 +1,155 @@
+import * as podmanDesktopApi from '@podman-desktop/api';
+import { readFile } from 'fs/promises';
+
+const DEFAULT_PORT = 10434;
+const SHOW_API_INFO_COMMAND = 'ai-lab.show-api-info';
+
+export class ApiServer implements Disposable {
+  #listener?: Server;
+
+  constructor(private extensionContext: podmanDesktopApi.ExtensionContext) {}
+
+  protected getListener(): Server | undefined {
+    return this.#listener;
+  }
+
+  async init(): Promise<void> {
+    const app = express();
+
+    const router = express.Router();
+    router.use(express.json());
+
+    // declare routes
+    router.get('/version', this.getVersion.bind(this));
+    app.use('/api', router);
+    app.use('/spec', this.getSpec.bind(this));
+
+    const server = http.createServer(app);
+    let listeningOn = DEFAULT_PORT;
+    server.on('listening', () => {
+      this.displayApiInfo(listeningOn);
+    });
+    server.on('error', () => {
+      getFreeRandomPort('0.0.0.0')
+        .then((randomPort: number) => {
+          console.warn(`port ${DEFAULT_PORT} in use, using ${randomPort} for API server`);
+          listeningOn = randomPort;
+          this.#listener = server.listen(randomPort);
+        })
+        .catch((e: unknown) => {
+          console.error('unable to get a free port for the api server', e);
+        });
+    });
+    this.#listener = server.listen(DEFAULT_PORT);
+  }
+
+  displayApiInfo(port: number): void {
+    const apiStatusBarItem = podmanDesktopApi.window.createStatusBarItem();
+    apiStatusBarItem.text = `AI Lab API listening on port ${port}`;
+    apiStatusBarItem.command = SHOW_API_INFO_COMMAND;
+    this.extensionContext.subscriptions.push(
+      podmanDesktopApi.commands.registerCommand(SHOW_API_INFO_COMMAND, async () => {
+        const address = `http://localhost:${port}`;
+        const result = await podmanDesktopApi.window.showInformationMessage(
+          `AI Lab API is listening on\n${address}`,
+          'OK',
+          `Copy`,
+        );
+        if (result === 'Copy') {
+          await podmanDesktopApi.env.clipboard.writeText(address);
+        }
+      }),
+      apiStatusBarItem,
+    );
+    apiStatusBarItem.show();
+  }
+
+  getSpecFile(): string {
+    // when plugin is installed, the file is placed in the plugin directory (~/.local/share/containers/podman-desktop/plugins/<pluginname>/)
+    const prodFile = path.join(__dirname, '../api/openapi.yaml');
+    if (existsSync(prodFile)) {
+      return prodFile;
+    }
+    // return dev file
+    return path.join(__dirname, '../../../api/openapi.yaml');
+  }
+
+  getPackageFile(): string {
+    // when plugin is installed, the file is placed in the plugin directory (~/.local/share/containers/podman-desktop/plugins/<pluginname>/)
+    const prodFile = path.join(__dirname, '../package.json');
+    if (existsSync(prodFile)) {
+      return prodFile;
+    }
+    // return dev file
+    return path.join(__dirname, '../../../package.json');
+  }
```

Comment:
looks like it should be externalized as it's a copy/paste

Reply:
Should be good now